### ISSUE TITLE
needless_lifetime is not emitted for async functions

### ISSUE THREAD

**Issue by Veetaha** [rust-lang/rust-clippy#5983]
I tried this code:

```rust
async fn foo<'a>() {}

fn main() {
    let _ = foo();
}
```

I expected to see a `needless_lifetime` warning from clippy.

Instead, no warnings were emitted (without the `async` modifier, the lint is emitted).

### Meta

- `cargo clippy -V`: `clippy 0.0.212 (04488af 2020-08-24)`
- `rustc -Vv`:
  ```
    rustc 1.46.0 (04488afe3 2020-08-24)
    binary: rustc
    commit-hash: 04488afe34512aa4c33566eb16d8c912a3ae04f9
    commit-date: 2020-08-24
    host: x86_64-unknown-linux-gnu
    release: 1.46.0
    LLVM version: 10.0
  ```

**Comment by ebroto** [comment]
Hey @Veetaha, thanks for opening the issue!

I think this is expected behavior, though I agree it's not apparent on first sight. According to the [async/await RFC](https://github.com/rust-lang/rfcs/blob/master/text/2394-async_await.md#lifetime-capture-in-the-anonymous-future), `All of the input lifetimes to this function are captured in the future returned by the async function`.

The return value of the function in your example is `impl Future<Output = ()> + 'a`, even though the lifetime is not actually used.

I'm gonna close the issue, please reopen it in case you find it necessary.

**Comment by ebroto** [comment]
Without knowing the details about the implementation, one could argue that in this case the lifetime should not be captured as it's not part of the function input parameters, so maybe an issue in [rust-lang/rust](https://github.com/rust-lang/rust) could be opened.

**Comment by Veetaha** [comment]
Hmm, this indeed seems inappropriate because the lifetime here doesn't pertain to any of the arguments.
Why do you think this should be an issue in [rust-lang/rust](https://github.com/rust-lang/rust). May there be some bug in `rustc`?

**Comment by ebroto** [comment]
Not sure if it could be considered a bug, but it would seem `rustc` is being too conservative given that the lifetime is not used in the arguments (just in the generic ones). But maybe this was an implementation decision?

On Clippy's side, I feel we can't make the decision that the lifetime is unused if it's actually used in the return type. For example this lint won't trigger in this case ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=97fdf0f0c595558535172901534faa5c)) in which the lifetime is not used in the input but just in the output (here `'a` would be normally spelled `'static`, but this is still valid code).

**Comment by CAD97** [comment]
An alternate lint that would catch it: detect input lifetimes (caller-chosen) that are not bound in any way. Typically, it'd be clearer to write that using the `'static` lifetime.

To make the lint more correct, check that the lifetime is used in a variant manner, such that callers can actually coerce from `'static` to a shorter lifetime.
